File: marvelHeroes/src/main/java/com/futureworkshops/marvelheroes/presentation/character/list/dagger/CharacterListModule.java
/*
 * Copyright (c) 2018 FutureWorkshops. All rights reserved.
 */

package com.futureworkshops.marvelheroes.presentation.character.list.dagger;

import com.futureworkshops.marvelheroes.data.network.rx.scheduler.SchedulersProvider;
import com.futureworkshops.marvelheroes.domain.navigator.Navigator;
import com.futureworkshops.marvelheroes.domain.repositories.character.MarvelCharacterRepository;
import com.futureworkshops.marvelheroes.presentation.character.list.CharacterListInteractor;
import com.futureworkshops.marvelheroes.presentation.character.list.CharacterListPresenter;
import com.futureworkshops.marvelheroes.presentation.character.list.CharactersListContract;
import com.futureworkshops.marvelheroes.presentation.character.list.view.CharacterListFragment;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;

/**
 * Created by stelian on 04/04/2018.
 */
@Module()
public abstract class CharacterListModule {
    
    @Provides
    static Navigator providesNavigator(CharacterListFragment fragment) {
        return new Navigator(fragment.getActivity());
    }
    
    @Binds
    abstract CharactersListContract.View providesContractView(CharacterListFragment fragment);
    
    @Provides
    static CharacterListInteractor providesInteractor(MarvelCharacterRepository newsRepository) {
        return new CharacterListInteractor(newsRepository);
    }
    
    @Provides
    static CharacterListPresenter providesPresenter(CharacterListInteractor interactor, Navigator navigator,
                                                    CharactersListContract.View view, SchedulersProvider schedulersProvider) {
        return new CharacterListPresenter(interactor, navigator, view, schedulersProvider);
    }
}

File: marvelheroeskotlin/src/main/java/com/futureworkshops/marvelheroeskotlin/presentation/character/list/CharacterListModule.kt
/*
 * Copyright (c) 2018 FutureWorkshops. All rights reserved.
 */

package com.futureworkshops.marvelheroeskotlin.presentation.character.list

import android.app.Activity
import com.futureworkshops.marvelheroeskotlin.domain.AppComponent
import com.futureworkshops.marvelheroeskotlin.domain.FragmentScope
import com.futureworkshops.marvelheroeskotlin.domain.navigator.Navigator
import com.futureworkshops.marvelheroeskotlin.domain.repository.MarvelCharacterRepository
import dagger.Component
import dagger.Module
import dagger.Provides
import io.reactivex.disposables.CompositeDisposable

/**
 * Created by dimitrios on 20/06/2018.
 * Copyright © 2018 - Future Workshops.
 * All rights reserved.
 */

@Module
class CharacterListModule(val fragment: CharacterListFragment) {
    
    @FragmentScope
    @Provides
    fun providesNavigator(): Navigator {
        return Navigator(fragment.activity as Activity)
    }
    
    @FragmentScope
    @Provides
    internal fun providesInteractor(newsRepository: MarvelCharacterRepository): CharacterListInteractor {
        return CharacterListInteractor(newsRepository)
    }
    
    @FragmentScope
    @Provides
    internal fun providesPresenter(interactor: CharacterListInteractor, navigator: Navigator, compositeDisposable: CompositeDisposable): CharacterListPresenter {
        return CharacterListPresenter(interactor, navigator, compositeDisposable)
    }
    
}

@FragmentScope
@Component(dependencies = [(AppComponent::class)], modules = [CharacterListModule::class])
interface CharacterListComponent {
    
    fun inject(fragment: CharacterListFragment)
    
}